akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,501 @@
+
+  final class InputState(signalDemand: AsyncCallback[Long]) {
+    private var untilNextDemandSignal = DemandThreshold
+
+    def onElement(): Unit = {
+      untilNextDemandSignal -= 1
+      if (untilNextDemandSignal == 0) {
+        untilNextDemandSignal = DemandThreshold
+        signalDemand.invoke(DemandThreshold)
+      }
+    }
+
+    def close(): Unit = signalDemand.invoke(MergeHub.Cancel)
+
+  }
+
+  final class MergedSourceLogic(_shape: Shape, producerCount: AtomicLong) extends GraphStageLogic(_shape) with OutHandler {
+    private val queue = new AbstractNodeQueue[Event] {}
+    @volatile private[this] var needWakeup = false
+    @volatile private[this] var shuttingDown = false
+
+    private[this] val demands = new java.util.HashMap[Long, InputState]
+    private[this] val wakeupCallback = getAsyncCallback[NotUsed]((_) ⇒
+      // We are only allowed to dequeue if we are not backpressured. See comment in tryProcessNext() for details.
+      if (isAvailable(out)) tryProcessNext(firstAttempt = true)
+    )
+
+    setHandler(out, this)
+
+    // Returns true when we have not consumed demand, false otherwise
+    private def onEvent(ev: Event): Boolean = ev match {
+      case Element(id, elem) ⇒
+        demands.get(id).onElement()
+        push(out, elem)
+        false
+      case Register(id, callback) ⇒
+        // TODO: Check for already existing state? Should not happen if works correctly though
+        demands.put(id, new InputState(callback))
+        true
+      case Deregister(id) ⇒
+        demands.remove(id)
+        true
+    }
+
+    override def onPull(): Unit = tryProcessNext(firstAttempt = true)
+
+    @tailrec private def tryProcessNext(firstAttempt: Boolean): Unit = {
+      val nextElem = queue.poll()
+
+      // That we dequeue elements from the queue when there is demand means that Register and Deregister messages
+      // might be delayed for arbitrary long. This is not a problem as Register is only interesting if it is followed
+      // by actual elements, which would be delayed anyway by the backpressure.
+      // Unregister is only used to keep the map growing too large, but otherwise it is not critical to process it
+      // timely. In fact, the only way the map could keep growing would mean that we dequeue Registers from the
+      // queue, but then we will eventually reach the Deregister message, too.
+      if (nextElem ne null) {
+        needWakeup = false
+        if (onEvent(nextElem)) tryProcessNext(firstAttempt = true)
+      } else {
+        needWakeup = true
+        // additional poll() to grab any elements that might missed the needWakeup
+        // and have been enqueued just after it
+        if (firstAttempt)
+          tryProcessNext(firstAttempt = false)
+      }
+    }
+
+    def isShuttingDown: Boolean = shuttingDown
+
+    // External API
+    def enqueue(ev: Event): Unit = {
+      queue.add(ev)
+      if (needWakeup) {
```

Comment:
Isn't this `wakeup` thing 
*exactly* the same desired behavior as the one which is modeled for actor scheduling?
(I say this because this implementation is drastically different and I trust few of these solutions before extensive, high-intensity testing)